### ISSUE TITLE
Add maybeUndefined type

### ISSUE THREAD

**Issue by xaviergonz** [mobxjs/mobx-state-tree#249]
Basically like the maybe type but for undefined instead of null
sugar for:
types.optional(types.union(type, types.literal(undefined)), undefined)

Why undefined instead of null? See here:

https://github.com/Microsoft/TypeScript/wiki/Coding-guidelines#null-and-undefined
Use undefined. Do not use null.

https://basarat.gitbooks.io/typescript/docs/tips/null.html
Null is Bad

**Comment by mattiamanzati** [comment]
Hello!
I am completely ok for this issue, the only issue is the naming, I'd avoid names with multiple cases (they are all lowercased intentionally), so can we name this `option`?
We already have `maybe` and `optional` is for omittable fields in snapshot.
Another option would be to make `optional` behave like this if no second argument is provided. But I don't like much this idea.

**Comment by xaviergonz** [comment]
then maybe add 'nullable'  as alias for maybe and then also add 'undefinable'?

**Comment by xaviergonz** [comment]
or add a first optional argument to maybe that can be either null or undefined and defaults to null if there's only one argument. so it would look like 
maybe(undefined, Todo) or maybe(null, Todo)

**Comment by mattiamanzati** [comment]
What about types.option(Todo, undefined)? Also @mweststrate is smarter than me at naming things ;)

**Comment by mweststrate** [comment]
Maybe we should just redefine `maybe` to 

```
union(optional(undefinedType, undefined), wrappedType)
```

and explicitly export

```
types.null
types.undefined
```

So we can in general recommend to use `undefined`, and `maybe` uses undefined, and `types.null` is just there in case you really need for example because that is what you get from the server.

(And also should remove `null` from code base while at it :-P ?)

**Comment by xaviergonz** [comment]
+1 that's what major versions are for I guess

**Comment by mattiamanzati** [comment]
Uhm, I think that if we move maybe from null to undefined we should at least provide a codemod or something that fixes that, but +1 for the rest :)

**Comment by chemitaxis** [comment]
Yes, we have sometimes null or undefined values from server, and I think that a codemod could be nice ;)

**Comment by mweststrate** [comment]
Add least `types.null` and `types.undefined` will be exposed explicitly in the next release, so It should become easy to change the behavior locally if needed. A `0.10` is a semantic breaking version, so I think it should be okayish to change the `maybe` type entirely? It is hard to predict what the impact would be on existing code bases

**Comment by mweststrate** [comment]
Hmm having second thoughts on using `undefined` as default value, given this:

```
> JSON.stringify({ a: undefined })
'{}'
> JSON.stringify({ a: null })
'{"a":null}'
```

Omitting a key because it has `undefined` value is a very tricky thing imho, it might not be acceptable by server validations, and might be very confusing when updating object, when trying to create a system that can receive patches etc. 

E.g. : `Object.assign({ x: 3 }, { x: undefined })` is something completely different from `Object.assign({ x: 3, {})`

**Comment by xaviergonz** [comment]
So then it is recommended to use null instead of undefined?

**Comment by mweststrate** [comment]
Yep

Op za 22 jul. 2017 22:42 schreef xaviergonz <notifications@github.com>:

**Comment by mondaychen** [comment]
Then what's the final decision? Do we keep `maybe`'s current behavior and export `types.undefined` and `types.null`?

**Comment by mweststrate** [comment]
Yep most likely 😊

Op ma 24 jul. 2017 16:36 schreef Monday Chen <notifications@github.com>:

**Comment by mweststrate** [comment]
Released 0.9.4, exposed `types.null` and `types.undefined`, but kept `maybe` as is. Also see https://twitter.com/mweststrate/status/888848172860264448
